server/proof_fetcher.go

```diff
@@ -1,53 +1,155 @@
-	// TODO(Martin2112): Implement the rehashing. Currently just fetches the nodes and ignores this
-	proofNodeIDs := make([]storage.NodeID, 0, len(proofNodeFetches))
+	proofNodes, err := dedupAndFetchNodes(tx, treeRevision, proofNodeFetches)
+	if err != nil {
+		return trillian.Proof{}, err
+	}
+
+	r := newRehasher()
+	for i, node := range proofNodes {
+		r.process(node, proofNodeFetches[i])
+	}
+
+	return r.rehashedProof(leafIndex)
+}
+
+// rehasher bundles the rehashing logic into a simple state machine
+type rehasher struct {
+	th         merkle.TreeHasher
+	rehashing  bool
+	rehashNode storage.Node
+	proof      []*trillian.Node
+	proofError error
+}
+
+// init must be called before the rehasher is used or reused
+func newRehasher() *rehasher {
+	return &rehasher{
+		// TODO(Martin2112): TreeHasher must be selected based on log config.
+		th: merkle.NewRFC6962TreeHasher(crypto.NewSHA256()),
+	}
+}
+
+func (r *rehasher) process(node storage.Node, fetch merkle.NodeFetch) {
+	switch {
+	case !r.rehashing && fetch.Rehash:
+		// Start of a rehashing chain
+		r.startRehashing(node)
 
-	for _, fetch := range proofNodeFetches {
-		proofNodeIDs = append(proofNodeIDs, fetch.NodeID)
+	case r.rehashing && !fetch.Rehash:
+		// End of a rehash chain, resulting in a rehashed proof node
+		r.endRehashing()
+		// And the current node needs to be added to the proof
+		r.emitNode(node)
+
+	case r.rehashing && fetch.Rehash:
+		// Continue with rehashing, update the node we're recomputing
+		r.rehashNode.Hash = r.th.HashChildren(node.Hash, r.rehashNode.Hash)
+
+	default:
+		// Not rehashing, just pass the node through
+		r.emitNode(node)
+	}
+}
+
+func (r *rehasher) emitNode(node storage.Node) {
+	idBytes, err := proto.Marshal(node.NodeID.AsProto())
+	if err != nil {
+		r.proofError = err
+	}
+	r.proof = append(r.proof, &trillian.Node{NodeId: idBytes, NodeHash: node.Hash, NodeRevision: node.NodeRevision})
+}
 
-		// TODO(Martin2112): Remove this when rehashing is implemented
-		if fetch.Rehash {
-			return trillian.Proof{}, errors.New("proof requires rehashing but it's not implemented yet")
+func (r *rehasher) startRehashing(node storage.Node) {
+	r.rehashNode = storage.Node{Hash: node.Hash}
+	r.rehashing = true
+}
+
+func (r *rehasher) endRehashing() {
+	if r.rehashing {
+		r.proof = append(r.proof, &trillian.Node{NodeHash: r.rehashNode.Hash})
+		r.rehashing = false
+	}
+}
+
+func (r *rehasher) rehashedProof(leafIndex int64) (trillian.Proof, error) {
+	r.endRehashing()
+	return trillian.Proof{LeafIndex: leafIndex, ProofNode: r.proof}, r.proofError
+}
+
+// dedupAndFetchNodes() removes duplicates from the set of fetches and then passes the result to
+// storage. After writng this code I realised I don't have a solid proof that fetches for Merkle
+// paths could involve duplicate nodes so it could be that this isn't ever useful. Further
+// thought is required.
```

Comment:
I don't have an actual proof though that this doesn't happen. For example if we needed to recompute multiple nodes on the RHS path but I suspect it doesn't. I'm going to remove the deduplication until it's proven to be needed.